### ISSUE TITLE
build SWIG with CMake

### ISSUE THREAD

**Issue by KrisThielemans** [swig/swig#1039]
It would be nice to be able to build SWIG with CMake.

I have put some more effort into this. See https://github.com/KrisThielemans/swig/tree/cmake

This is a continuation of work discussed in 
https://sourceforge.net/p/swig/mailman/message/34804657/

I've now merged master into the cmake branch, and did a few more fixes. I can currently build Swig on Linux and Visual Studio (tested with 2015). The python test-suite runs as well except for one test (with a new test in `insert_directive`).

There's still a few work-arounds here, but not a lot. Maybe somebody else can have a look at his.

I haven't created a PR as my previous changes to CCache will likely break the normal build. I guess we could just undo them.

**Comment by xantares** [comment]
hey,
I had put some work into that branch, but after experimentng (it's a lot of work to get all the tests in all languages running, it's much slower that the makefiles, not counting the cost of maintaining it if primary devs only use the autotools part) I think we'd rather provide a very basic cmake support (ie just build/install which is already available) so that our visual studio users are able to compile it easily (and possibly linux users too).
It would be easy for me to cut the tests part.

**Comment by xantares** [comment]
see #1040

**Comment by KrisThielemans** [comment]
thanks.

getting it to build swig only isn't so hard, but I guess my version is a bit more advanced by using stuff from Bjoern etc (but I didn't check).

Running the test-sutie is indeed very slow, but a normal "build all" doesn't do that, so most people wouldn't be affected. But I agree that maintaining all the test-suites is a fair amount of work. I guess I'd still vote for having one or two basic ones (like Python), such that people can still check if it works or not. If the SWIG developers want to avoid the maintainers burden, a half-way solution would be to update the CMake branch with larger support, but put only basic things in master.

**Comment by ojwb** [comment]
The test suite is already slower than I'd like - I'm definitely not keen on making it slower.  We want to encourage people providing patches to actually run the tests first, and we want CI builds to finish sooner.

**Comment by xantares** [comment]
@sethrj It's not meant to run the test suite, only building swig. The thing is that if you want to write every rule in a platform-independent manner it gets significantly slower than the existing scripts (I tried).

**Comment by sethrj** [comment]
The CMake should handle the platform independence, and I've never seen a CMake script slower than its autoconf equivalent. Is the makefile slow? Using the "ninja" output of CMake might be worthwhile.

**Comment by xantares** [comment]
The makefile is slow yes. ninja does not change much, only runs the makefiles in parallel by default.

**Comment by xantares** [comment]
@wsfulton I think you can close this now

**Comment by KrisThielemans** [comment]
Fine for me. I would suggest creating another issue then that says "add running of test-suite to CMake", pointing to initial work that I did at https://github.com/KrisThielemans/swig/tree/cmake (which is now hopelessly out of date)

**Comment by xantares** [comment]
maintaining the test-suite for both build systems is a burden, I wont take it that far

**Comment by KrisThielemans** [comment]
for sure, but I think have the "issue" means that at least people are aware of it that it hasn't been done yet.

**Comment by xantares** [comment]
i dont think it should be done

**Comment by KrisThielemans** [comment]
Which means that the test-suite will never be run on non-`configure` systems. Anyway,  as I certainly don't have the time to pick my effort up, feel free to close.

**Comment by xantares** [comment]
only you or @wsfulton can do that

**Comment by wsfulton** [comment]
CMake will only ever replace the autotools if someone puts in a massive effort to replicate everything it does and it is no slower.

For the record the https://github.com/KrisThielemans/swig/tree/cmake branch was built on top of the cmake branch in https://github.com/swig/swig. If anyone ever looks at porting the test-suite to CMake, https://github.com/KrisThielemans/swig/tree/cmake could be used. I've deleted https://github.com/swig/swig/tree/cmake (6ab477cfa1a958e6a0000300b8f5b419378494ad) in Github from which https://github.com/KrisThielemans/swig/tree/cmake was branched.
